solutions/CF Garland.mdx

```diff
@@ -0,0 +1,120 @@
+
+[Official Editorial](https://codeforces.com/blog/entry/52449)
+
+## Alternative Solution (Dynamic Programming)
+
+**Time Complexity:** $\mathcal{O}(n^2k + q)$
+
+Since we are performing up to $200,000$ queries, we must have some way to quickly precompute our answer in constant time.
+
+We can use dynamic programming to cache such answers. 
+
+Define $\texttt{dp}[i][j][k]$ to be the best possible score of the first $i$ characters, with $j$ amount painted of character $k$. Either two things can happen when we add an additional character.
+
+1. We append the character and paint an additional square.
+
+<center>
+
+$$\texttt{dp}[i+1][j+1][k] = \max(\texttt{dp}[i+1][j+1][k], \texttt{dp}[i][j][k] + 1)$$
+
+</center>
+
+2. We append the character, but since it is the same as our previous character we get to "save" a paint.
+
+<center>
+
+$$\texttt{dp}[i+1][j][k] = \max(\texttt{dp}[i+1][j][k], \texttt{dp}[i][j][k] + 1)$$
+
+</center>
+
+Since this only gives us the best possible score that ends with length $i$, we can just take the current best and translate it using $\texttt{dp}[i+1][j][k] = \max(\texttt{dp}[i+1][j][k], \texttt{dp}[i][j][k])$
+
+Now, we can query the max painting score for any substring containing the first character. However, the this is not required, we only have to query the optimal paintings with size $n$ (the full string) with $\texttt{dp}[n][m_i][c_i]$
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h> // see /general/running-code-locally
+using namespace std;
+
+using ll = long long;
+
+using vi = vector<int>;
+#define pb push_back
+#define all(x) begin(x), end(x)
+#define sz(x) (int)(x).size()
+
+using pi = pair<int,int>;
+#define f first
+#define s second
+#define mp make_pair
+
+void setIO(string name = "") {
+	cin.tie(0)->sync_with_stdio(0); // see /general/fast-io
+	if (sz(name)) {
+		freopen((name+".in").c_str(), "r", stdin); // see /general/io
+		freopen((name+".out").c_str(), "w", stdout);
+	}
+}
+
+// i = length of string
+// j = amount painting
+// k = character
+
+int dp[1501][1501][26];
+int S[1501];
+
+int main() {
+	setIO();
+
+	int n; cin >> n;
+	for(int i = 0; i < n; i++) {
+		char a; cin >> a;
+		S[i] = (int)(a - 'a');
+	}
+
+	for(int k = 0; k < 26; k++) { // lower case characters
+		for(int i = 0; i < n; i++) {
+			for(int j = 0; j < n; j++) {
+				// appending a character
+				dp[i+1][j+1][k] = max(dp[i+1][j+1][k], dp[i][j][k] + 1);
+
+				// if we can "save"
+				if(S[i] == k)
+					dp[i+1][j][k] = max(dp[i+1][j][k], dp[i][j][k] + 1);
+			}
+		}
+	}
+
+	// take the current best and translate it
+	// dp[i+1][j][k] = max(dp[i+1][j][k], dp[i][j][k]);
+
+	for(int k = 0; k < 26; k++)
+		for(int i = 0; i < n; i++)
+			for(int j = 0; j < n; j++)
+				dp[i+1][j][k] = max(dp[i+1][j][k], dp[i][j][k]);
+
+	int q;
+	cin >> q;
+
+	for(int i = 0; i < q; i++) {
+		int a; char b;
+		cin >> a >> b;
+		b -= 'a';
+
+		cout << dp[n][a][(int)b] << '\n';
+	}
+}
+```
+
+</CPPSection>
+
+</LanguageSection>
```

Comment:
```suggestion
---
id: cf-garland
source: CF
title: An impassioned circulation of affection
author: Neo Wang
---

[Official Editorial](https://codeforces.com/blog/entry/52449)

## Alternative Solution (Dynamic Programming)

**Time Complexity:** $\mathcal{O}(n^2k + q)$

Since we are performing up to $200,000$ queries, we must have some way to quickly precompute our answer in constant time.

We can use dynamic programming to cache such answers. 

Define $\texttt{dp}[i][j][k]$ to be the best possible score of the first $i$ characters, with $j$ amount painted of character $k$. Either two things can happen when we add an additional character.

1. We append the character and paint an additional square. <center>
$$\texttt{dp}[i+1][j+1][k] = \max(\texttt{dp}[i+1][j+1][k], \texttt{dp}[i][j][k] + 1)$$</center>
2. We append the character, but since it is the same as our previous character we get to "save" a paint.<center>$$\texttt{dp}[i+1][j][k] = \max(\texttt{dp}[i+1][j][k], \texttt{dp}[i][j][k] + 1)$$</center>


Since this only gives us the best possible score that ends with length $i$, we can just take the current best and translate it using $\texttt{dp}[i+1][j][k] = \max(\texttt{dp}[i+1][j][k], \texttt{dp}[i][j][k])$

Now, we can query the max painting score for any substring containing the first character. However, the this is not required, we only have to query the optimal paintings with size $n$ (the full string) with $\texttt{dp}[n][m_i][c_i]$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h> // see /general/running-code-locally
using namespace std;
using ll = long long;
using vi = vector<int>;
#define pb push_back
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()
using pi = pair<int,int>;
#define f first
#define s second
#define mp make_pair
void setIO(string name = "") {
	cin.tie(0)->sync_with_stdio(0); // see /general/fast-io
	if (sz(name)) {
		freopen((name+".in").c_str(), "r", stdin); // see /general/io
		freopen((name+".out").c_str(), "w", stdout);
	}
}
// i = length of string
// j = amount painting
// k = character
int dp[1501][1501][26];
int S[1501];
int main() {
	setIO();
	int n; cin >> n;
	for(int i = 0; i < n; i++) {
		char a; cin >> a;
		S[i] = (int)(a - 'a');
	}
	for(int k = 0; k < 26; k++) { // lower case characters
		for(int i = 0; i < n; i++) {
			for(int j = 0; j < n; j++) {
				// appending a character
				dp[i+1][j+1][k] = max(dp[i+1][j+1][k], dp[i][j][k] + 1);
				// if we can "save"
				if(S[i] == k)
					dp[i+1][j][k] = max(dp[i+1][j][k], dp[i][j][k] + 1);
			}
		}
	}
	// take the current best and translate it
	// dp[i+1][j][k] = max(dp[i+1][j][k], dp[i][j][k]);
	for(int k = 0; k < 26; k++)
		for(int i = 0; i < n; i++)
			for(int j = 0; j < n; j++)
				dp[i+1][j][k] = max(dp[i+1][j][k], dp[i][j][k]);
	int q;
	cin >> q;
	for(int i = 0; i < q; i++) {
		int a; char b;
		cin >> a >> b;
		b -= 'a';
		cout << dp[n][a][(int)b] << '\n';
	}
}
```

</CPPSection>

</LanguageSection>
```